### PR TITLE
Update inputs on device in `tt_transformers`

### PR DESCRIPTION
**Pull request by rdraskicTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/21291

Pending CI runs to see to see final improvements, but preliminary numbers show increase for llama 8B DP=4 on TG from 44 t/s/u to 51 t/s/u.

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/16836881780) CI passes
- [x] [TG demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/16836893216) CI passes
- [x] [vllm nightly](https://github.com/tenstorrent/tt-metal/actions/runs/16836565733)
- [x] [llama 8B DP=4](https://github.com/tenstorrent/tt-shield/actions/runs/16836920483)

### CODE REVIEW

**Review by gwangTT — APPROVED** [review]
LGTM. Thanks for the optimization work!

File: models/tt_transformers/tt/model.py [hunk 1]
```diff
@@ -339,10 +340,17 @@ def ttnn_decode_forward(
...
+        elif not self.args.is_galaxy:
             # Send output logits to DRAM so L1 is not reserved for ttnn tracing and can be used by subsequent operations
-            if not self.args.is_galaxy:
-                tt_logits = ttnn.to_memory_config(tt_logits, ttnn.DRAM_MEMORY_CONFIG)
+            tt_logits = ttnn.to_memory_config(tt_logits, ttnn.DRAM_MEMORY_CONFIG)
+
+        if argmax_on_device and update_on_device:
+            # Update device tensors for the next iteration
+            ttnn.plus_one(current_pos)
+            ttnn.plus_one(rot_mat_idxs)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/model.py]: Should this only do +1 for the current_pos rows which are not -1 (padded users)? Not sure how it currently works for Llama70B on Galaxy if they always do +1.

Comment (rdraskicTT) [reply models/tt_transformers/tt/model.py]: That's a good point, I'm not sure why this works either. Alternative is to have a more complicated formula here but I'm not sure if it's worth the trouble.

Comment (skhorasganiTT) [reply models/tt_transformers/tt/model.py]: @cglagovich we use -1 in paged_update_cache and paged_sdpa to determine users to skip right?

Comment (skhorasganiTT) [reply models/tt_transformers/tt/model.py]: @rdraskicTT I think it should be fixed otherwise we might overwrite the caches for other users with the paged updates.

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -207,12 +208,22 @@ def decode_forward_text(
         enable_trace=True,
         read_from_device=True,
         sampling_params: SamplingParams = None,  # Should be None if not greedy decoding / sampling on device.
+        update_on_device=False,
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py]: why is it default here False, and in llama-tg it's True?

File: models/tt_transformers/demo/simple_text_demo.py [hunk 3]
```diff
@@ -773,6 +773,7 @@ def test_demo_text(
                 page_table=page_table,
                 kv_cache=tt_kv_cache,
                 sampling_params=device_sampling_params,
+                update_on_device=device_sampling_params is not None,
```
Comment (ppetrovicTT) [inline models/tt_transformers/demo/simple_text_demo.py]: what is the use-case to have update_on_device False and sampling_params working?
can we deduce that update_on_device is True always if sampling_params are provided?
(i'm thinking about removing it from the API if it can be deduced inside the function).

also, `update_on_device` name doesn't really explain what it does, especially at this level (`decode_forward_text` high level function)

File: models/tt_transformers/tt/generator.py [hunk 4]
```diff
@@ -363,25 +366,38 @@ def _easy_trace_text(
         page_table=None,
         kv_cache=None,
         argmax_on_device=False,
+        update_on_device=False,
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py]: for my taste, these 3 arguments (argmax_on_device, update_on_device, reset_inputs) should be combined into a struct. and do we really need all of them? or some can be deduced from others?

they all impact the same area of the execution and i feel they pollute the API on a high level

File: models/tt_transformers/tt/generator.py (hunk 2 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: My reasoning was that some models are still being run without sampling on device in tt-inference-server and I didn't want to break those. llama-tg has both enabled so it doesn't matter.

File: models/tt_transformers/demo/simple_text_demo.py (hunk 3 above)
Comment (rdraskicTT) [reply models/tt_transformers/demo/simple_text_demo.py]: In principle no, we can remove it. I put a flag because I didn't want to activate this for all models at the moment but I can remove it if you think I should activate it by default when sampling on device is active. I don't have a better name, can you suggest an alternative to `update_on_device`?

File: models/tt_transformers/tt/generator.py (hunk 4 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: I agree, we can combine these into some sort of generator_params so it will be easier to remove them once they become fully supported. Should we address it in this PR or do a followup?

File: models/tt_transformers/tt/generator.py (hunk 2 above)
Comment (ppetrovicTT) [reply models/tt_transformers/tt/generator.py]: why would they break? i thought this would benefit perf for all the models in tt_transformers, was i wrong?

File: models/tt_transformers/demo/simple_text_demo.py (hunk 3 above)
Comment (ppetrovicTT) [reply models/tt_transformers/demo/simple_text_demo.py]: hint for the name: what are you updating?

as it is now, looked from the high-level (`decode_forward_text`), seems like the model should be updated on device.

File: models/tt_transformers/tt/generator.py (hunk 2 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: You are right, I started with this assumption but the way the code looks now it no longer makes sense. `update_on_device` is already coupled to `sampling_params` so I'll just remove it.

File: models/tt_transformers/tt/generator.py (hunk 4 above)
Comment (ppetrovicTT) [reply models/tt_transformers/tt/generator.py]: in this PR i would like us to make sure we need all these arguments.

in case you want to make this struct, `generator_params` still sounds quite vague. their actual purpose should be reflected in the name.

Comment (mtairum) [reply models/tt_transformers/tt/generator.py]: We also should move from doing argmax on host. This was there from the time we had a bug when running multicore argmax for batch >1 if I recall correctly.

Plus add the top-K sampling support to TT-Transformers now that we have the op working.
This would simplify all these calls.

File: models/tt_transformers/tt/model.py (hunk 1 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/model.py]: @skhorasganiTT I fixed the update, could you take another look?

File: models/tt_transformers/tt/model.py [hunk 5]
```diff
@@ -320,7 +320,8 @@ def _update_decode_inputs_device(self, current_pos, rot_mat_idxs):
         result = ttnn.typecast(result, ttnn.uint32)
         ttnn.copy(ttnn.to_layout(result, layout=ttnn.ROW_MAJOR_LAYOUT), rot_mat_idxs)
 
-        return current_pos, rot_mat_idxs
+        # Update input tokens with sampled tokens for the next iteration
+        ttnn.copy(tokens_out.reshape(tokens.shape), tokens)
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/model.py]: why did you move this inside the function?
the previous version seems cleaner to me

Comment (rdraskicTT) [reply models/tt_transformers/tt/model.py]: I thought it would be confusing if `_update_decode_inputs_device` didn't update all input tensors. But if you prefer the previous version I can revert.

Comment (ppetrovicTT) [reply models/tt_transformers/tt/model.py]: then maybe the name doesn't fit

Comment (rdraskicTT) [reply models/tt_transformers/tt/model.py]: Reverted and renamed, let me know if it makes sense now.

**Review by ppetrovicTT — APPROVED** [review]

File: models/tt_transformers/tt/generator.py [hunk 6]
```diff
@@ -213,6 +214,11 @@ def decode_forward_text(
         ), "Currently only supporting greedy decoding (temperature=0) on device"
         argmax_on_device = sampling_params is not None and sampling_params.temperature == 0
 
+        reset_inputs = not argmax_on_device
+        if self.prev_page_table is None or torch.any(self.prev_page_table != page_table).item():
+            reset_inputs = True
+            self.prev_page_table = page_table
+
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Can you move this inside the `if enable_trace` statement below?

File: models/tt_transformers/tt/generator.py [hunk 7]
```diff
@@ -327,30 +332,11 @@ def _capture_trace_text(
...
-        for i in range(self.data_parallel):
-            to_device.append(
-                copy_host_to_device(
-                    host_tensors=host_inputs[i],
-                    device_tensors=device_inputs[i],
-                )
-            )
-        device_inputs = to_device
         for i, trace_id in trace_ids.items():
             ttnn.execute_trace(self.model_args[i].mesh_device, trace_id, cq_id=0, blocking=False)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: I think you might as well move this into `_easy_trace_text` as well and delete `_decode_forward_trace_text` now since this function doesn't do much anymore.

File: models/tt_transformers/tt/generator.py [hunk 8]
```diff
@@ -314,10 +320,9 @@ def _capture_trace_text(
             trace_id = ttnn.begin_trace_capture(self.model_args[i].mesh_device, cq_id=0)
             trace_ids[i] = trace_id
             user_kv_cache = kv_cache[i] if kv_cache is not None else None
-            transformed_inputs = self.model[i].transform_decode_inputs_device(*(device_inputs[i]))
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:317]: I think all the function changes (reset_inputs is optional, doesn't need to be in this PR) you've done for the text models should also be done for the vision model functions (_capture_trace, etc.) to keep them consistent.

File: models/tt_transformers/tt/model.py [hunk 9]
```diff
@@ -300,11 +297,34 @@ def ttnn_prefill_forward(
             kv_cache=kv_cache,
         )
 
+    def _increment_decode_positions_device(self, current_pos, rot_mat_idxs):
+        current_pos_tiled = ttnn.to_layout(current_pos, layout=ttnn.TILE_LAYOUT)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/model.py:302]: Why does it need to be converted to tile layout? Is that a restriction of the ne or where ops? If so, maybe mention in comment.
Same question for rot mat idxs below.

File: models/tt_transformers/tt/model.py [hunk 10]
```diff
@@ -300,11 +297,34 @@ def ttnn_prefill_forward(
...
+        rot_mat_idxs_signed = ttnn.typecast(ttnn.to_layout(rot_mat_idxs, layout=ttnn.TILE_LAYOUT), ttnn.int32)
+        # Update only active positions (rot_mat_idxs_signed != 0)
+        predicate = ttnn.nez(rot_mat_idxs_signed)
+        result = ttnn.where(
+            predicate,
+            ttnn.add(rot_mat_idxs_signed, 1),
+            rot_mat_idxs_signed,
+        )
+        result = ttnn.typecast(result, ttnn.uint32)
+        ttnn.copy(ttnn.to_layout(result, layout=ttnn.ROW_MAJOR_LAYOUT), rot_mat_idxs)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/model.py]: I think you should just do plus_one for rot_idxs since 0 might be a valid position so it should be incremented.

File: models/tt_transformers/tt/model.py [hunk 11]
```diff
@@ -300,11 +297,34 @@ def ttnn_prefill_forward(
             kv_cache=kv_cache,
         )
 
+    def _increment_decode_positions_device(self, current_pos, rot_mat_idxs):
+        current_pos_tiled = ttnn.to_layout(current_pos, layout=ttnn.TILE_LAYOUT)
+        # Update only active positions (current_pos != -1)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/model.py:303]: FYI @mtairum @sraizada-tt this same logic may need to be applied for Llama-Galaxy. I'm not sure how you're currently not running into accuracy issues from cache overwrites due to incrementing the -1 positions.

File: models/tt_transformers/tt/model.py (hunk 9 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/model.py:302]: Yes, currently a restriction of the ops, will add a comment.

File: models/tt_transformers/tt/generator.py (hunk 8 above)
Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py:317]: I agree, however, I would leave this for the followup as well. I looked into it and it will be a big change in terms of number of lines so I prefer to do it in a separate PR.

**Review by skhorasganiTT — APPROVED** [review]
